lib/connection_pool.rb

```diff
@@ -53,22 +53,28 @@ def initialize(options = {}, &block)
   end
 
   def with(options = {})
+    success = false # hoisted
     conn = checkout(options)
     begin
-      yield conn
+      (yield conn).tap do
+        success = true # means the connection wasn't interrupted
+      end
     ensure
-      checkin
+      if success
+        # everything is roses, we can safely check the connection back in
+        checkin
+      else
+        @available.discard!(pop_connection)
```

Comment:
@zmoazeni at work, we encountered connection-in-use errors from Mysql2 for a long time. Only after adding an explicit mutex lock as per my comment at https://gist.github.com/josevalim/470808#comment-1377343 did the problem disappear.